src/ILVerify/src/SimpleTypeSystemContext.cs

```diff
@@ -3,118 +3,44 @@
 
-        Dictionary<string, EcmaModule> _modules = new Dictionary<string, EcmaModule>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<PEReader, EcmaModule> _modulesCache = new Dictionary<PEReader, EcmaModule>();
 
-        class ModuleData
+        public SimpleTypeSystemContext(IResolver resolver)
         {
-            public string Path;
-        }
-        Dictionary<EcmaModule, ModuleData> _moduleData = new Dictionary<EcmaModule, ModuleData>();
-
-        public SimpleTypeSystemContext()
-        {
-        }
-
-        public IDictionary<string, string> InputFilePaths
-        {
-            get;
-            set;
-        }
-
-        public IDictionary<string, string> ReferenceFilePaths
-        {
-            get;
-            set;
+            _resolver = resolver;
         }
 
         public override ModuleDesc ResolveAssembly(AssemblyName name, bool throwIfNotFound = true)
         {
-            return GetModuleForSimpleName(name.Name);
-        }
-
-        public EcmaModule GetModuleForSimpleName(string simpleName)
-        {
-            EcmaModule existingModule;
-            if (_modules.TryGetValue(simpleName, out existingModule))
-                return existingModule;
-
-            return CreateModuleForSimpleName(simpleName);
-        }
-
-        private EcmaModule CreateModuleForSimpleName(string simpleName)
-        {
-            string filePath;
-            if (!InputFilePaths.TryGetValue(simpleName, out filePath))
+            PEReader peReader = _resolver.Resolve(name);
+            if (peReader == null && throwIfNotFound)
             {
-                if (!ReferenceFilePaths.TryGetValue(simpleName, out filePath))
-                    throw new CommandLineException("Assembly not found: " + simpleName);
+                throw new VerifierException("Assembly or module not found: " + name.Name);
             }
 
-            PEReader peReader = new PEReader(File.OpenRead(filePath));
-            EcmaModule module = EcmaModule.Create(this, peReader);
-
-            MetadataReader metadataReader = module.MetadataReader;
-            string actualSimpleName = metadataReader.GetString(metadataReader.GetAssemblyDefinition().Name);
-            if (!actualSimpleName.Equals(simpleName, StringComparison.OrdinalIgnoreCase))
-                throw new CommandLineException("Assembly name does not match filename " + filePath);
-
-            _modules.Add(simpleName, module);
-
-            ModuleData moduleData = new ModuleData() { Path = filePath };
-            _moduleData.Add(module, moduleData);
-
-            return module;
-        }
-
-        public EcmaModule GetModuleFromPath(string filePath)
-        {
-            // This is called once for every assembly that should be verified, so linear search is acceptable.
-            foreach (KeyValuePair<EcmaModule, ModuleData> entry in _moduleData)
+            string actualSimpleName = peReader.GetSimpleName();
```

Comment:
This will create MetadataReader and throw it away immediately. The GetModule method a few lines below will create the MetadataReader again.

It would be better to move this validation to after the GetModule call (and delete the GetSimpleName method because of it should not be needed anymore).